### PR TITLE
optimize ServiceRequest methods

### PR DESCRIPTION
**Pull request by fakeshadow**
## PR Type

Refactor

## PR Checklist
Check your PR fulfills the following:

- [x] Documentation comments have been added / updated.
- [x] Format code with the latest stable rustfmt

## Overview

Optimize `ServiceRequest::from_parts` by remove duplicate ref counter check.
Optimize `ServiceRequest::from_request` by ignore weak pointer check.
Add comment on `HttpRequest` for the non weak pionter nature of `Rc<HttpRequestInner>`

### CODE REVIEW

File: src/service.rs [hunk 1]
```diff
@@ -73,11 +73,12 @@ impl ServiceRequest {
         mut req: HttpRequest,
         pl: Payload,
     ) -> Result<Self, (HttpRequest, Payload)> {
-        if Rc::strong_count(&req.0) == 1 && Rc::weak_count(&req.0) == 0 {
```
Comment (robjtede) [inline src/service.rs:76]: keep the weak count as a debug assertion here and on from_request

File: src/request.rs [hunk 2]
```diff
@@ -17,6 +17,9 @@ use crate::rmap::ResourceMap;
 #[derive(Clone)]
 /// An HTTP Request
 pub struct HttpRequest(pub(crate) Rc<HttpRequestInner>);
+// *. Rc<HttpRequestInner> is used exclusively and NO Weak<HttpRequestInner>
+// is allowed anywhere in the code. Weak pointer is purposely ignored when
+// doing Rc's ref counter check.
```
Comment (robjtede) [inline src/request.rs]: make this a proper doc comment on HttpRequestInner

**Review by robjtede — CHANGES_REQUESTED** [review]

File: src/service.rs (hunk 1 above)
Comment (fakeshadow) [reply src/service.rs:76]: `Rc::get_mut` only work on this condition. There is no need to double check.

Comment (fakeshadow) [reply src/service.rs:76]: `Rc::get_mut` only work on this condition. `from_request` indeed worth a debug assert

File: src/request.rs (hunk 2 above)
Comment (fakeshadow) [reply src/request.rs]: HttpRequestInner type does not know the outer Rc wrapper. 
I feel the comment should be at HttpRequest level. Could be worth make HttpRequest a struct with named field for a clear view of doc.

Comment (robjtede) [reply src/request.rs]: yep i like this too

**Review by robjtede — APPROVED** [review]

**Pull request merged**